### ISSUE TITLE
Add default locations

### ISSUE THREAD

**Issue by JacobEvelyn** [JacobEvelyn/friends#152]
The idea here is that I can say `set home NYC` or `set home yesterday NYC` and then a special activity is added that signifies I have set/changed my home location.

Then, commands like `list activities --in NYC` will automatically include activities _after that point_ that do not have a location specified. This is useful for when you're in a location for a longer period of time (e.g. a 1-year sabbatical) and you don't want to have to add the location name to every single activity.

**Comment by shen-sat** [comment]
Hi @JacobEvelyn! I'm a new-ish coder (I've been devloper for a year and a bit now), and I've always wanted to contribute to open source. I searched on Github, found this project and have completely fallen in love with it - such a cool idea!

I want to work on this issue as my first ever open source contribution, but before I do, I noticed a problem with the running the tests on my machine. I've fixed it, but in order to show you the fix, would I open a pull request?

**Comment by JacobEvelyn** [comment]
Hi @shen-sat! So sorry for the delay!

First of all, welcome to coding and open source! 👋

Thanks so much for finding and fixing a bug! A pull request would be great (and if you have any questions or problems with that let me know and I'm happy to help). I'm also assigning this ticket to you so you can officially work on it once you've fixed the issue with tests. How does that sound?

**Comment by shen-sat** [comment]
Hi @JacobEvelyn! So sorry for the late reply, I almost missed this message - I didn't get a notification in my emails for some reason...I'll check my settings 👍

I'm so happy to be assigned my first issue woop! 🙌 I will be taking this project to a local Ruby meetup next week, so I can get someone more senior to pair with me on it. I'll therefore start work on this next week 👨‍💻

PS please ignore my tweet to you today, I sent it thinking you might not have seen my Github message 😅

**Comment by JacobEvelyn** [comment]
Sounds great, @shen-sat! Good luck at the meetup, and feel free to reach out if you have any questions!

**Comment by shen-sat** [comment]
Hey @JacobEvelyn, I've started looking at this issue. I managed to pair with someone to understand the general app structure, but I'm going to attempt writing tests and building the `set home` functionality on my own (until I get stuck, haha).

As I understand it, there are two requirements:
1.  User can set a home location
2. User can list activities by home location

I want to tackle the first requirement first. For this, I'm thinking of creating a new header in the friends.md file (`### Home`) to store the home location.

Then there are two user journeys I can think of:
1. User sets home for the first time (the app will add a location under `### Home`)
2. User sets a new home when one already exists (the app will overwrite the existing location under `### Home` with the new location)

Does that sound ok? I'll be using your existing tests to TDD this functionality 💪 🙂

**Comment by JacobEvelyn** [comment]
Hey @shen-sat! Thanks so much for picking this up, and great questions.

I've actually been thinking about this a little bit differently. Consider this activity history (in chronological order, which is the opposite of the ordering in `friends.md`):

```
# (Living in London at start.)
- 2019-01-01: Hung out with **Banksy** at home.
- 2019-02-01: Moved to _Atlanta_.
- 2019-03-01: Went back to _London_ to visit **Banksy**.
- 2019-04-01: **Banksy** came to visit me!
```

Because you can move home locations, what I really would like is for the `list activities --in` flag to understand the "home" concept, like:

```
$ friends list activities --in London
2019-03-01: Went back to London to visit Banksy.
2019-01-01: Hung out with Banksy at home.
```
```
$ friends list activities --in Atlanta
2019-03-01: Banksy came to visit me!
2019-02-01: Moved to Atlanta.
```

So I think storing in `### Activities` (or `### Notes`??) might make more sense, because then we naturally have a timeline that we can use.

Open questions I have:

1. Should it actually be a `set home` command, or just a special activity name?
2. Should this command create the location if it doesn't already exist? (I think probably yes.)
3. Where's the best place to store this? (I'm leaning `### Activities` over `### Notes` or `### Home`, but am open to arguments!)
4. Should the home location impact results of `friends list notes --in <location>`? What are some use cases for filtering notes by location?

I hope that all makes sense! Would love to hear your thoughts. ❤️

**Comment by shen-sat** [comment]
Hey @JacobEvelyn, thanks for the detailed user examples, they help me understand the requirements a lot. I like the idea of using the existing user feed to add extra functionality, rather than adding an extra `### Home` section

My thoughts on your questions: 
1. Should it actually be a set home command, or just a special activity name? _I think so - as a user, I'd like to use a small set of commands to do lots of different things. There is already a `set` command for setting other things, so `friends set home <location>` would make intuitive sense. Plus my memory is bad, so fewer commands = 👍🏾 in any app, haha_
2. Should this command create the location if it doesn't already exist? (I think probably yes.) _Agree!_
3. Where's the best place to store this? (I'm leaning ### Activities over ### Notes or ### Home, but am open to arguments!) _It will be easier to store a location under `### Home`. But using the existing feed to store home location intrigues me :) (I'll expand on this below)_
4. Should the home location impact results of friends list notes --in <location>? What are some use cases for filtering notes by location? _I think so: I go to a place in NYC with a friend, I record the name of a shop I went to in my notes. Later I can't recall the name of the shop but remember it is in NYC so I filter notes by NYC and then I see the name of the shop in one of the entries. And if my home is in Paris, it will be useful for a `friends list notes --in NYC` to automatically filter out all the entries that are either explicity tagged with `Paris` or have no location (because the app will assume entries with no location are from Paris)_

Re: point 3 - with the example you gave:
```
# (Living in London at start.)
- 2019-01-01: Hung out with **Banksy** at home.
- 2019-02-01: Moved to _Atlanta_.
- 2019-03-01: Went back to _London_ to visit **Banksy**.
- 2019-04-01: **Banksy** came to visit me!
``` 
can we assume that:
- on the 2019-02-01, the user entered `friend set home Atlanta`
- when the user enters `friends list activities --in Atlanta`, the app does the following things:
 - finds `Moved to _Atlanta_` line in the `### Activities` section
 - collects the lines in the rest of the section until it either finds another `Moved to` line or reaches end of the section
- then filters out lines that are tagged with another location
- displays the lines that are left

As always, happy to discuss :)

**Comment by JacobEvelyn** [comment]
Thanks for all the discussion, @shen-sat, and sorry for the slow response. I think we're in agreement that:

1. `friends set home Atlanta` is the right command to use for this.
2. That command should create the location if it doesn't already exist.
3. We should include this when filtering notes by location, as well as activities (your example with the shop name is compelling to me).
4. The code will simply find a special line, rather than read data from a new `### Home` section.

After thinking about it a bit more, I'm wondering:

1. Is `Moved to _Atlanta_` under `### Activities` trying to be a bit too clever? Would it make more sense to just have it under `### Notes` instead as `Set home to _Atlanta_.`? Thoughts?
2. Is it useful to leverage the existing date-parsing code we have that lets us do things like `friends add note 3 days ago` or `friends add activity 2019-01-01`, so we can set the move date with `set home` to an arbitrary date in the past? (This is the one thing that's making me second-guess the new command instead of just having it be a special note you add with `friends add note`.)

Lastly, from an implementation perspective, the code reads the entire file before displaying anything to the screen for any command, so I think the easiest approach might be to have a new instance variable, `@home_locations`, that stores something like a hash of date range to location name:

```ruby
{
  (#<Date: 2019-01-01>..#<Date: 2019-01-31>) => "London",
  (#<Date: 2019-02-01>..#<Date: 2019-10-01>) => "Atlanta"
}
```

And then when filtering we can iterate through each until we find a range that matches. It's not the _most_ efficient approach, especially with lots of home locations (which I assume most users won't have), and I'm totally open to alternate/more performant ideas. Thoughts?

**Comment by shen-sat** [comment]
Hey @JacobEvelyn, 

Completely agree with leveraging the date parsing, that will make it easier to create the `@home)locations` variable you mentioned and populate it with a hash of locations and move dates.

I also agree with maybe not listing `Moved to _Atlanta_` under `### Activities`. However, I also don’t want to move it under `## Notes` either, and I’m leaning back towards creating a separate section for logging the user’s home location. Sorry for the flip-flopping on this! I’m wondering quite a lot about separation of concerns/responsibilities. Let me expand:

Looking through the code, what I love about your app atm is how neatly components have one responsibility. `### Activities` section lists activities with your friends. `### Notes` lists notes to do with your friends. I kind of don’t want to pollute these lists with events that are personal to the user and have functionality outside of logging the user’s friendships. Setting a user’s home is a way of filtering activities and notes in the future by location, rather than something that explicitly logs something to do with a user’s friends.

From a user perspective, when I list activities or notes, do I want to see a `Moved home _Atlanta_`? Maybe, but this would detract from what I see as Friends’ unique core functionality - to track and display your relationship with your friends, (and not to track and display personal events).

I’m therefore back in favour of having a separate section to log a user’s move to a new home 😄 🏠 . We could even call the section `### Personal` to make it open to logging other features you may want to add in the future that are to do with the user and not with their friendships.

What do you think?

**Comment by JacobEvelyn** [comment]
Thanks for all the thoughts! I'm not sure I have strong preferences either way, though I don't really know what other things would be in `### Personal` so my vote would be to just use `### Homes` for now, like:

```
### Activities
- 2019-01-01: Blah blah

### Notes
- 2019-02-01: Blah blah

### Homes
- 2019-03-01: Atlanta
- 2019-01-01: London

### Friends
- Banksy

### Locations
- Atlanta
- London
```

How does that sound?

**Comment by JacobEvelyn** [comment]
Hmm actually now I'm having second thoughts. One thing I've noticed is that even when I'm not _moving permanently_, if I'm traveling it gets _really_ annoying to add the location to every activity I do on my travels, like:

```
### Activities
- 2019-01-03: One final lunch with **Banksy** in _London_ before leaving. What a great trip!
- 2019-01-03: **Banksy** and I visited the Royal Botanic Gardens, Kew (_London_) and had a great conversation.
- 2019-01-02: Had a lovely dinner in _London_ with **Banksy** and **Queen Elizabeth II**.
- 2019-01-02: After the museum, **Queen Elizabeth II** took me out to tea (_London_).
- 2019-01-02: Went to the Tate Modern (_London_) with **Queen Elizabeth II** in the morning.
- 2019-01-01: Arrived in _London_ and went straight to Trafalgar Square with **Banksy** to check that off the travel list!
```

I'm constantly forgetting to add them, or putting them in parentheticals which feels awkward.

However, `friends`'s philosophy of [**intelligence**](https://github.com/JacobEvelyn/friends#overview) often leads to behaviors that are difficult to describe or understand but "just work" most of the time and feel like magic when they do (like how it almost always "just knows" which friend you mean when you have multiple friends with the same name). I'd rather have a more complicated implementation for simpler usage that serves the most common cases well.

From that perspective, I actually want to propose a different solution to both the "home location" and the "travel" problem: any **activity** that includes the phrase `from _<location>_ to _<location>_` will automatically be treated as changing the "home" location until the next time that phrase appears in a different activity. With this solution, all I'd need to start my trip would be something like `friends add activity Flew from Atlanta to London for my big trip!` and then something similar at the end (which I find I tend to use anyway), and then the rest would just work.

I know this is a totally different direction but for me as a user this would be significantly nicer. Obviously the downside is we'd need to document it and make it clear what's happening. Would love to hear your thoughts!

**Comment by shen-sat** [comment]
I love the fact we are going deep on this issue - I’m a dev, but I’ve been fascinated by product, and this feels like a juicy product challenge :)

I agree that user shouldn’t unneccesarily have to add their location all the time. So the concept of a default location that can change whether the user is at home or on holiday makes a lot of sense. I'd like to set this as the core objective for this issue.

I think we have two different views on how to implement this. I’ll lay out my thinking here, but I’m very happy to implement your solution - after all, whilst I love to give my opinion, I am ultimately here to code :)

Friends is an app to track relationships with friends. The examples used in the readme for the `### Activities` and `### Notes` sections are all to do with this vision. I really like this core objective - to me, it's simplicity/pureness is a strength of this application.

Once there is a feature to start logging events like moving home, or going on holiday, I feel like this would detract from the core objective. For me, activities/notes like this aren’t nesserily to do with relationships with friends anymore, it’s more to do with personal journaling.

However, I completely understand why we would want to track these events somehow - as I said at the top (and as you mentioned with the concept of intelligence), user’s shouldn’t have to repeatedly type in the same location every time while at home/on holiday. The Friends app should pick this up naturally and handle it as necessary.

To solve this, I think a config command like `friends set home [location]` feels intuitive. There are already commands for setting `friends` and `locations`, which are in and of themselves not notes or activities. They are to me, config commands. And once the config for home location is in place, the user can type in lines like `Banksy came to visit me!` and Friends acts intelligently - it picks up the home location automatically and acts as appropriate.

I see the home command following the same process as the other 'config' commands: first the user does a little config work. But then the Friends app pays the user back by handling input to do with home location intelligently.

The end result is what was already present: Activity logs only have friends activity. Same with notes. But now the user has to do less work re: writing locations for every note/activity.

Let me know what you think @JacobEvelyn - and I’ll start work based on your decision 👍🏾

**Comment by JacobEvelyn** [comment]
Sorry for the delay, @shen-sat. I've been thinking about this a lot and keep going back and forth. (Definitely overthinking it! 😬)

I think one of the things tripping me up is the word "home." A command like `friends set home` feels like something to only use when you permanently move, not something for a one-week vacation. Perhaps I'd feel more open to that approach with a different name?

You're right that there is already a `friends set` command, but there's only one and it doesn't do a whole lot. Honestly I've never loved that command name... would something like this make sense instead:

```
friends move LOCATION # Moves your location, a.k.a. the feature added by this issue.
friends move LOCATION DATE # Moves your location for a given date, also added by this issue.
friends move FRIEND LOCATION # The new version of the old "set" command.
```

I'm hesitant to store the location history in yet another place in the file, but one alternative would be to store them in your `### Locations:` list, like:

```
### Locations:
- Atlanta [2019-01-01 to 2019-01-04, 2019-01-06 to present]
- London [2019-01-04 to 2019-01-06]
```

it might be a bit trickier to implement (and I guess rows—especially for your home location—could get really long), but I like that it matches the pattern for how we store friends with details, and still keeps the option of location nicknames (something I've been [wanting to add](https://github.com/JacobEvelyn/friends/issues/242)) with the same syntax as friend nicknames on the table.

On the other hand, this is making me realize that a big missing feature in all of these approaches is your home location. I live in Atlanta and don't tag most activities I do there with `_Atlanta_`, but if I move away it would be nice to be able to have them all under that location—_even though while living in Atlanta I traveled to other places on vacation._ This feature is making me once again think that maybe the `from _A_ to _B_` (or something similar) in `### Activities:` is the way to go.

Lastly, I hear your point about the core functionality being around tracking friends, and agree with the sentiment around staying close to that functionality. But there's no reason an activity needs to have a friend in it from a technical perspective, and it seems like it might make things generally easier to go that route (for one thing, the date parsing is free!). So I think I'm leaning towards that approach.

Sorry for the long, drawn-out discussion. If you want me to stop bikeshedding and just pick something I'm happy to. But I'm also happy to keep discussing with you, and would love to hear your thoughts!

**Comment by shen-sat** [comment]
Hey @JacobEvelyn, I think we are in agreement! 🙌🏾

Completely agree with the `friends move LOCATION` command you've suggested. And I also agree with how you want to store it.

I also agree that it'd be nice to have Friends 'know' that your location has changed when you are travelling (with a command like `from _A_ to _B_`). I'm not 100% sure on the way to implement this, but it would be a very cool feature.

Can I suggest something? As this is my first (proper) OS contribution, would it be possible to split this issue into two issues? The first would be to tackle the `friends move LOCATION` commands you listed above (which I feel confident about implemeneting), and the second would be how to implement the `from _A_ to _B_` command (which may/may not require more discussion)?

I'm off on holiday now and won't be back for 2 weeks - I'm switching my phone and laptop off for a digital detox (I've always wanted to do one :) ). I'd hate to lose my chance to contribute though, so would it be possible to keep this issue for me until I'm back?

**Comment by JacobEvelyn** [comment]
Hope your digital detox is going well, @shen-sat! I've been giving this a bunch more thought and I've come to a decision—I'd like to proceed with the following:

- When an activity includes the pattern `to _LOCATION_`, future activities count as being "in" that location until the next `to _LOCATION_` pattern.
    - **Rationale:**
        1. This ticket originally derived from a common pain point I experience as a regular user of `friends` for almost five years now. I've found that I naturally tend to include those "travel activities" anyway, so from a UX perspective this would require essentially no behavior change for these added benefits.
        2. By using activities for this rather than a separate date record elsewhere in `friends.md`, we naturally gain sub-day granularity (if someone travels three times in a day, activities between those movements can still be correctly categorized by location). This attribute would not be present in other solutions.
        3. By using activities for this we can take advantage of existing date-parsing behavior here.
        4. I've seen how adding this "magic extraction" from activities in #190 dramatically improved `friends`'s usability, and I think we can replicate some of that magic here.
    - **Implementation details:**
        1. There is no default location for activities until the first time this `to _LOCATION_` pattern is seen.
        2. Since we read the file line-by-line and thus in reverse-chronological order, one way to implement this would be to add an `@implicit_location` instance variable to `Activity`, and store activities in an array as we parse them until we encounter a `to _LOCATION_` line, at which point we set the `@implicit_location` for each of those activities, clear the array, and repeat.
            - In the future we can always optimize this by (for example) only performing it when the `friends` command actually needs this data, e.g. `friends list activities --in LOCATION`. For now though I think it's fine to just always do it.
        3. When a new activity matching the `to _LOCATION_` pattern is added, a message like `Default location set from START_DATE to FUTURE_DATE/present: _LOCATION_` is printed to `STDOUT`. This will **improve the discoverability of this feature.**
- Notes will also use these locations in filtering, just not with sub-day granularity.
    - **NOTE: While I think this is nice behavior, I'm willing to follow a "wait and see if it's really useful" strategy and hold off on implementing it, because I personally don't use many notes (and when I do they're rarely tied to the location I'm in) and I'm not sure how big of a need this is among users.
    - **Rationale:** See item 4 in https://github.com/JacobEvelyn/friends/issues/152#issuecomment-536203483.
    - **Implementation details:**
        1. There is no default location for notes until the first time this `to _LOCATION_` pattern is seen.
        2. Because notes are stored separately from activities, we have no way of achieving sub-day granularity here. Notes may therefore have multiple implicit locations if the user has multiple "move" activities on the same day.
        3. Since we read the file line-by-line and thus in reverse-chronological order, one way to implement this would be when we encounter a `to _LOCATION_` pattern in an activity we add to a global data structure tracking these locations.
            - In the future we can always optimize this by (for example) only constructing this data structure when the `friends` command actually needs this data, e.g. `friends list activities --in LOCATION`. For now though I think it's fine to just always do it.
- Default locations will not be explicitly stored elsewhere in the `friends.md` file.
    - **Rationale:** Any solution that does this runs some or all of the following risks (depending on implementation):
        1. Duplicating data already elsewhere in the file (for instance, dates in `### Activities:` or locations in `### Locations:`).
        2. Being unable to provide sub-day granularity.
        3. Reducing the readability of the file with extremely wide lines containing all date ranges for a location.
- A new command (`friends move`) is added which also writes to the `### Activities:` section. See https://github.com/JacobEvelyn/friends/issues/152#issuecomment-539735035.
- `friends set` is changed to `friends move` as described in https://github.com/JacobEvelyn/friends/issues/152#issuecomment-539735035.

**Comment by shen-sat** [comment]
Sounds good to me! I'll have a closer look at your implementation suggestion and let you know if I have any questions 👍🏾

**Comment by shen-sat** [comment]
Hey @JacobEvelyn sorry for the late response! I'm blaming jet lag 😪

I agree with the outline of your implementation - the idea to create an instance variable `@implicit_location` in Activity and set it across multiple Activity objects as the `friends.md` file is parsed makes sense.

I'm getting a bit stuck on exactly _where_ the setting of @implicit_location should take place though, and I'd like to go through my thinking if that's ok?

If I'm correct, the current parsing of lines happens as follows:

1. Introvert object is created
2. `friends.md` is read 
3. When a line in a section (for example, ###Activities) is read:
    a. Validation is done via regex
    b. If validation is ok, new Activity object is created
    c. This new Activity object is then saved within `@activities` instance variable in Introvert object

The key line where stage 3 happens is (`Introvert.rb, line:712`):
`instance_variable_get("@#{stage.id}") << stage.klass.deserialize(line)`

In the new implementation, what should happen at this line? 

At the moment we add Activity objects one-by-one to `@activities`, but I assume that in the new implementation these would be added to a temporary array until a `to _LOCATION_` line is parsed (at which point we'd do a 'set and flush': we'd set the `@implicit_location` on all the Activity objects in the temp array and then save them one by one to `@activities` and then clear the temp array).

Whose responsibility should it be to detect a when a `to _LOCATION_` line is being parsed? My first thought was this detection should somehow sit within the Activity class. But if an Activity object detects that a `to _LOCATION_` line is being parsed, how could it set the `@implicit_location` on all previous Activity objects within the temporary array? This feels like something the Introvert class can do, but then the Introvert class would be responsible for detecting a `to _LOCATION_` line, which doesn't make sense in terms of responsibility 🥴

I hope that makes sense! In essence my confusion comes around responsibility. If you're up for a video call I'd be super up for chatting to you and maybe explaining the above in real-time - I appreciate I might not have laid out my thinking very clearly! Otherwise I'm happy to reach out to someone at work to pair with them on this - either way is good for me :)

**Comment by shen-sat** [comment]
Hey @JacobEvelyn, I managed to pair with someone at work today and we managed to clear my above confusion - please ignore my questions above! I'll get on with the implementation and let you know how I get on 👍🏾

**Comment by JacobEvelyn** [comment]
Okay! Glad to hear it, and let me know if there's anything else I can help with.

**Comment by JacobEvelyn** [comment]
Thanks so much, @shen-sat ! This change has been released in v0.47 (run `friends update` to get it). Thanks for contributing!

EDIT: Whoops, I just realized that the change looks for `Moved to _location_` but I think we'd actually wanted it to just be `to _location_` (to allow things like "Flew to Hawaii for vacation" to work). I'm happy to make that change quickly unless you want to!

**Comment by shen-sat** [comment]
No worries!

> EDIT: Whoops, I just realized that the change looks for Moved to _location_ but I think we'd actually wanted it to just be to _location_ (to allow things like "Flew to Hawaii for vacation" to work). I'm happy to make that change quickly unless you want to!

Ah this is something I thought about - I used the more specific line 'moved to _location_' so that we wouldn't set implcit location for lines like "I went to Marie's Diner". What do you reckon?

**Comment by JacobEvelyn** [comment]
Hmm, that's an interesting thought. I've personally been using locations more for geographic areas/cities/etc. than specific buildings. I assume that's how most users are using them as well but I can't say for certain. Do you have locations like `_Marie's Diner_`?

**Comment by shen-sat** [comment]
`Marie's Diner` is used in a lot of tests - when I saw it, I thought locations in the context of the app could be geographic areas or specific spots like restaurants, museums, etc. But if not, then I'm happy to change `moved to _location_` to `to _location_`. But I think we should also lose `Marie's Diner` form the codebase as it can cause confusion over the definition of a location. What do you reckon @JacobEvelyn?

**Comment by JacobEvelyn** [comment]
I think that makes sense to me. Do you want to work on that, or should I?

Two other follow-on things that I forgot to mention on the last PR:

1. When an activity that sets a new implicit location is added, I think we should print a message. Something like `Default location set to: "#{location}"`
2. We should add a description of this feature to the README.

**Comment by shen-sat** [comment]
Makes sense and I'm happy to make the changes 👍 What's the best practice here - should I re-open the pull request, or make a new one?

**Comment by JacobEvelyn** [comment]
A new PR would be great! Thanks so much! Also, I can't figure out why you're not showing up on the [Contributors page](https://github.com/JacobEvelyn/friends/graphs/contributors). Any ideas? Perhaps something in your privacy settings?
